packages/loopring_v3/contracts/test/AmmPool.sol

```diff
@@ -360,111 +400,340 @@ contract AmmPool is IBlockReceiver {
+        // This also pulls the AMM balances onchain.
+        processAmmUpdates(ctx, true);
+
+        // Process all pool transactions
+        for (uint n = 0; n < poolTransactions.length; n++) {
+            PoolTransaction memory poolTx = poolTransactions[n];
+            if (poolTx.txType == PoolTransactionType.JOIN) {
+                PoolJoin memory join = abi.decode(poolTx.data, (PoolJoin));
+                processJoin(ctx, join, poolTx.signature);
+            } else if (poolTx.txType == PoolTransactionType.EXIT) {
+                PoolExit memory exit = abi.decode(poolTx.data, (PoolExit));
+                processExit(ctx, exit, poolTx.signature);
+            }
+        }
+
+        // Deposit/Withdraw to/from the AMM account when necessary
+        for (uint i = 0; i < ctx.tokens.length; i++) {
+            if (ctx.ammBalances[i] > ctx.ammBalancesInAccount[i]) {
+                uint amount = ctx.ammBalances[i] - ctx.ammBalancesInAccount[i];
+                processDeposit(ctx, ctx.tokens[i], amount);
+            } else if (ctx.ammBalancesInAccount[i] > ctx.ammBalances[i]) {
+                uint amount = ctx.ammBalancesInAccount[i] - ctx.ammBalances[i];
+                processWithdrawal(ctx, ctx.tokens[i], amount);
+            }
+        }
+
+        // The ending AMM updates
+        processAmmUpdates(ctx, false);
+
+        emit QueueItemsProcessed(poolTransactions.length);
+        return ctx.numTransactionsConsumed;
+    }
+
+    function totalSupply()
+        public
+        view
+        returns (uint256)
+    {
+        return poolSupply;
+    }
+
+    function mint(address owner, uint amount)
+        internal
+    {
+        poolSupply = poolSupply.add(amount);
+        poolBalance[owner] = poolBalance[owner].add(amount);
+    }
+
+    function burn(address owner, uint amount)
+        internal
+    {
+        poolSupply = poolSupply.sub(amount);
+        poolBalance[owner] = poolBalance[owner].sub(amount);
+    }
+
+    function processAmmUpdates(
+        Context memory ctx,
+        bool   start
+        )
+        internal
+    {
+        for (uint i = 0; i < ctx.tokens.length; i++) {
             // Check that the AMM update in the block matches the expected update
-            AmmUpdateTransaction.AmmUpdate memory update = _block.readAmmUpdate(txIdx++);
+            AmmUpdateTransaction.AmmUpdate memory update = ctx._block.readAmmUpdate(ctx.txIdx++);
             require(update.owner == address(this), "INVALID_TX_DATA");
             require(update.accountID == accountID, "INVALID_TX_DATA");
-            require(update.tokenID == tokens[i].tokenID, "INVALID_TX_DATA");
+            require(update.tokenID == ctx.tokens[i].tokenID, "INVALID_TX_DATA");
             require(update.feeBips == feeBips, "INVALID_TX_DATA");
-            require(update.tokenWeight == tokens[i].weight, "INVALID_TX_DATA");
+            require(update.tokenWeight == (start ? 0 : ctx.tokens[i].weight), "INVALID_TX_DATA");
```

Comment:
Not necessary because the nonce doesn't really matter for transactions authorized by the contract (can only be used once immediately). The relayer is free to choose the nonce/storageID (which needs to be valid, otherwise it can't be included in a block).

For transactions done by the user we I just made the storageIDs used by the transfers something that's signed by the user directly. This way we don't have to store anything onchain either, replay protection is done on layer 2.